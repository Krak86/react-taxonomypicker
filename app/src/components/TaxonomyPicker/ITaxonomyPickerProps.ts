export interface ITaxonomyPickerProps {
    displayName?: string;
    name: string;
    multi: boolean;
    termSetGuid: string;
    termSetName: string;
    termSetCountMax?: number;
};
