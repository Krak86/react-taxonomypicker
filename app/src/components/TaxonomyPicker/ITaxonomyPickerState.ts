export interface ITaxonomyPickerState {
    errors: string[];
    options: Object[];
    value: any;
    disabled: boolean;
    termSetCount: number;
    asyncLoad: boolean;
};
