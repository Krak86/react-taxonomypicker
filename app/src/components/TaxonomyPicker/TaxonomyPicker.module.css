.label {
  font-family: 'Segoe UI WestEuropean','Segoe UI',-apple-system,BlinkMacSystemFont,Roboto,'Helvetica Neue',sans-serif;
  text-align: left;
  font-size: 16px;
  display: block;
  color: white;
  margin-top: 10px;
  padding: 5px 0;
  -webkit-font-smoothing: antialiased;
  box-shadow: none;
  box-sizing: border-box;
}
