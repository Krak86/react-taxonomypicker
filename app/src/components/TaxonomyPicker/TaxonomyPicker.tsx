import * as React from "react";
import { Async, Select } from "react-select/lib/Select";
import TaxonomyAPI from "../../utils/MockAPI/SP.Taxonomy";

/* tslint:disable:no-var-requires */
const styles: any = require("./TaxonomyPicker.module.css");
/* tslint:enable:no-var-requires */

import { ITaxonomyPickerProps } from "./ITaxonomyPickerProps";
import { ITaxonomyPickerState } from "./ITaxonomyPickerState";

class TaxonomyPicker extends React.Component<ITaxonomyPickerProps, ITaxonomyPickerState> {

    // Default Props values
    public static defaultProps: ITaxonomyPickerProps = {
        multi: false,
        name: "",
        termSetCountMax: 350,
        termSetGuid: "",
        termSetName: "",
    };

    constructor(props: any, context: any) {
        super(props, context);

        // Default State values
        this.state = {
            asyncLoad: false,
            disabled: false,
            errors: [],
            options: [],
            termSetCount: 0,
            value: null
        };
    }

    // Initial Async Loading here. Only in Container Components
    public componentDidMount() {
        const { termSetGuid, termSetName, termSetCountMax } = this.props;
        TaxonomyAPI.getTermSetCount(termSetGuid, termSetName)
            .then((termSetCount: number) => {
                if (termSetCount > termSetCountMax) {
                    this.setState({ asyncLoad: true, termSetCount });
                } else {
                    this.getAllTerms()
                        .then((options: any) => {
                            this.setState({ options, termSetCount });
                        })
                        .catch((reason: any) => {
                            this.setState({ errors: [...this.state.errors, reason] });
                        });
                }
            })
            .catch((reason: any) => {
                this.setState({ errors: [...this.state.errors, reason] });
            });
    }

    public getSelectControl(async: boolean, loadOptions?: any, minimumInput?: number) {
        if (async) {
            return (<Select.Async
                isLoading={!this.state.options}
                backspaceRemoves={false}
                name={this.props.name}
                simpleValue={false}
                placeholder="Type here to search..."
                loadOptions={this._asyncLoadOptions}
                minimumInput={1}
                multi={this.props.multi}
                ref={this.props.name}
                onChange={this.handleSelectChange}
                options={this.state.options}
                value={this.state.value}
            />);
        } else {
            return (<Select
                isLoading={!this.state.options}
                backspaceRemoves={false}
                name={this.props.name}
                simpleValue={false}
                placeholder="Type here to search..."
                multi={this.props.multi}
                ref={this.props.name}
                onChange={this.handleSelectChange}
                options={this.state.options}
                value={this.state.value}
            />);
        }

    }

    public render() {
        const { asyncLoad } = this.state;
        return (
            <div key={this.props.name} className={styles.container} name={this.props.name}>
                <label className={styles.label} htmlFor={this.props.name}>{this.props.displayName}</label>
                {this.getSelectControl(asyncLoad, this._asyncLoadOptions, 1)}
                {this.state.errors.length > 0 ? this.renderErrorMessage() : null}
            </div>
        );
    }

    private _asyncLoadOptions = (input) => {
        return this.getSearchTerms(input)
            .then((options: any) => {
                this.setState({ options });
                return { options };
            })
            .catch((reason: any) => {
                this.setState({ errors: [...this.state.errors, reason] });
            });
    }

    private getSearchTerms(input: string) {
        const termFetcher = TaxonomyAPI.getSearchTermsByText(this.props.termSetGuid, this.props.termSetName, input);
        return termFetcher;
    }

    private getAllTerms() {
        const termFetcher = TaxonomyAPI.getAllTermsByTermSet(this.props.termSetGuid, this.props.termSetName, false);
        return termFetcher;
    }

    private renderErrorMessage() {
        return (
            <div>
                {this.state.errors}
            </div>
        );
    }

    private handleSelectChange = (value: any) => {
        this.setState({ value });
    }
}

export default TaxonomyPicker;
