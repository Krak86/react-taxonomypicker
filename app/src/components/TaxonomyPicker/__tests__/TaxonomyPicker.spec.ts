import testComponentHelper from "../../../../../test/unit/helpers/ComponentHelper";
import TaxonomyPicker from "../TaxonomyPicker";

describe("<TaxonomyPicker />", () => {
    const renderComponent = testComponentHelper(TaxonomyPicker);

    describe("@renders", () => {
        it("in default state", () => {
            expect(renderComponent().getHtml())
                .toMatchSnapshot();
        });
    });
});
