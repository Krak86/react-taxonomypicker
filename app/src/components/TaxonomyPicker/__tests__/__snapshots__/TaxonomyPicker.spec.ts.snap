// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TaxonomyPicker /> @renders in default state 1`] = `
<div
  className="container"
  key="Taxononomy_Picker_Name"
  title="Taxononomy_Picker_Name"
>
  <Select
    arrowRenderer={[Function]}
    autosize={true}
    backspaceRemoves={false}
    backspaceToRemoveMessage="Press backspace to remove {label}"
    clearAllText="Clear all"
    clearRenderer={[Function]}
    clearValueText="Clear value"
    clearable={true}
    closeOnSelect={true}
    defaultOptions={null}
    defaultValue={null}
    deleteRemoves={true}
    delimiter=","
    disabled={false}
    escapeClearsValue={true}
    filterOption={[Function]}
    filterOptions={[Function]}
    ignoreAccents={false}
    ignoreCase={true}
    inputProps={Object {}}
    isLoading={false}
    joinValues={false}
    labelKey="label"
    logErrorsConsole={false}
    logErrorsDiv={false}
    matchPos="any"
    matchProp="any"
    menuBuffer={0}
    menuRenderer={[Function]}
    multi={true}
    name="Taxononomy_Picker_Name"
    noResultsText="No results found"
    onBlurResetsInput={true}
    onChange={[Function]}
    onCloseResetsInput={true}
    onPickerChange={null}
    onSelectResetsInput={true}
    openOnClick={true}
    optionComponent={[Function]}
    optionRenderer={[Function]}
    options={Array []}
    pageSize={5}
    placeholder="Type here to search..."
    removeSelected={true}
    required={false}
    rtl={false}
    scrollMenuIntoView={true}
    searchable={true}
    showPath={false}
    simpleValue={false}
    tabSelectsValue={true}
    termSetAllTermsCacheExpiresMin={1440}
    termSetCountCacheExpiresMin={10080}
    termSetCountMaxSwapToAsync={300}
    termSetGuid={null}
    termSetName={null}
    trimFilter={true}
    value={null}
    valueComponent={[Function]}
    valueKey="value"
  >
     
     
  </Select>
</div>
`;
